tools/serde-templated-derive/README.md

```diff
@@ -0,0 +1,199 @@
+* `Vec<T>` -> `Vec<Templated<T>>`
+* `BTreeMap<K, V>` -> `BTreeMap<Templated<K>, Templated<V>>`
+* `BTreeSet<T>` -> `BTreeSet<Templated<T>>`
+* `BinaryHeap<T>` -> `BinaryHeap<Templated<T>>`
+* `HashMap<K, V>` -> `HashMap<Templated<K>, Templated<V>>`
+* `HashSet<T>` -> `HashSet<Templated<T>>`
+* `LinkedList<T>` -> `LinkedList<Templated<T>>`
+* `VecDeque<T>` -> `VecDeque<Templated<T>>`
+
+The generated `serde_templated::TemplateOf` implementation will treat each field of the generated
+struct as having implemented `TemplateOf<Target = T>` where `T` is the type of that same field on
+the parent struct.
+
+These behaviors can be overridden with various macro options.
+
+### Macro Options
+
+#### `derive`
+By default, the generated `Templated` struct only implements `Serialize` and `Deserialize`.
+You can instruct the macro to derive additional traits like so:
+
+```rust
+#[derive(Templated)]
+#[templated(derive(Default))]
+struct MyStruct {
+    value: u64,
+}
+```
+
+#### `forward_attrs`
+Allows you to copy an attribute from the "source" struct to the "templated" struct.
+
+```rust
+#[derive(Templated)]
+#[templated(forward_attrs(serde))]
+#[serde(rename_all = "kebab-case")] // The resulting TemplatedMyStruct will also have this attr
+struct MyStruct {
+    value: u64,
+
+    #[templated(forward_attrs(serde))]  // Can also be specified for fields
+    #[serde(rename = "different-name")]
+    forwarded_field: String,
+}
+```
+
+#### `templated_attrs`
+This allows you to define arbitrary attributes on the resulting struct, even if they aren't on the
+parent struct.
+
+Due to a syntax parsing limitation, these should be given as strings.
+
+```rust
+#[derive(Templated)]
+#[templated(templated_attrs = "#[derive(Debug)]")]
+struct Example {
+    #[templated(templated_attrs = "#[serde(alias = \"templated-value\")]")] // Can also be used for fields
+    value: Option<u64>,
+}
+```
+
+#### `skip_serde_derive`
+By default, the generated struct will derive implementations of `Serialize` and `Deserialize`.
+This behavior can be disabled with the `skip_serde_derive` attribute.
+
+```rust
+#[derive(Templated)]
+#[templated(skip_serde_derive)]
+struct Example {
+    // ...
+}
+```
+
+#### `skip`
+Can only be used for fields.
+
+This causes a sub-field of the generated struct to not be "templated" and instead to parse the
+given field using the same behavior as the parent struct.
+
+```rust
+#[derive(Templated)]
+struct Example {
+    #[templated(skip)] // `TemplatedExample` will also have this type as a `String`
+    value: String,
+}
+```
+
+#### `template_as`
+Can only be used for fields.
+
+Instread of using the default behavior to generate the templated type, fields will be templated
+using the provided type.
+
+Due to a limitation of syntax parsing for macros, this is given as a string.
+
+```rust
+#[derive(Templated)]
+struct Example {
+    #[templated(template_as = "Option<Templated<String>>")]
+    value: String,
+}
```

Comment:
Not really a doc question, but how would this work in practice if the template ended up with a `None` and the struct requires something in the `value` field?

Reply:
This would result in code that doesn't compile, unless you also provide a `render_with` attribute to `templated` that tells it how to do the conversion back from the template to the parent.